src/HealthGPS/static_linear_model.cpp

```diff
@@ -83,81 +105,173 @@ void StaticLinearModel::update_risk_factors(RuntimeContext &context) {
     for (size_t i = 0; i < names_.size(); i++) {
 
         // Initialise residual.
         auto residual_name = core::Identifier{names_[i].to_string() + "_residual"};
-        double residual_new = residuals[i];
-        person.risk_factors[residual_name] = residual_new;
+        double residual = residuals[i];
+        person.risk_factors[residual_name] = residual;
 
         // Initialise risk factor.
-        double factor_new = linear_factors[i] + residual_new * stddev_[i];
         double expected = get_risk_factor_expected().at(person.gender, names_[i]).at(person.age);
-        factor_new = expected * inverse_box_cox(factor_new, lambda_[i]);
-        person.risk_factors[names_[i]] = factor_new;
+        double factor = linear[i] + residual * stddev_[i];
+        factor = expected * inverse_box_cox(factor, lambda_[i]);
+        person.risk_factors[names_[i]] = factor;
     }
 }
 
 void StaticLinearModel::update_factors(Person &person, Random &random) const {
 
-    // Approximate risk factor values with linear models.
-    auto linear_factors = compute_linear_models(person);
-
     // Correlated residual sampling.
-    auto residuals = compute_residuals(random);
+    auto residuals = compute_residuals(random, cholesky_);
+
+    // Approximate risk factor values with linear models.
+    auto linear = compute_linear_models(person, models_);
 
+    // Update residuals and risk factors (should exist).
     for (size_t i = 0; i < names_.size(); i++) {
 
         // Update residual.
         auto residual_name = core::Identifier{names_[i].to_string() + "_residual"};
-        double residual_new = residuals[i] * info_speed_;
-        residual_new += sqrt(1.0 - info_speed_ * info_speed_) * person.risk_factors[residual_name];
-        person.risk_factors[residual_name] = residual_new;
+        double residual_old = person.risk_factors.at(residual_name);
+        double residual = residuals[i] * info_speed_;
+        residual += sqrt(1.0 - info_speed_ * info_speed_) * residual_old;
+        person.risk_factors.at(residual_name) = residual;
 
         // Update risk factor.
-        double factor_new = linear_factors[i] + residual_new * stddev_[i];
         double expected = get_risk_factor_expected().at(person.gender, names_[i]).at(person.age);
-        factor_new = expected * inverse_box_cox(factor_new, lambda_[i]);
-        person.risk_factors[names_[i]] = factor_new;
+        double factor = linear[i] + residual * stddev_[i];
+        factor = expected * inverse_box_cox(factor, lambda_[i]);
+        person.risk_factors.at(names_[i]) = factor;
+    }
+}
+
+void StaticLinearModel::initialise_policies(Person &person, Random &random, bool intervene) const {
+    // NOTE: we need to keey baseline and intervention scenarios RNG in sync.
+
+    // Correlated residual sampling.
+    auto residuals = compute_residuals(random, policy_cholesky_);
+
+    // No-op if intervention scenario.
+    if (!intervene) {
+        return;
+    }
```

Comment:
Residuals are going to be used only it intervene is true, so why not checking this before? Actually, why running this function at all if intervene is false? I would make the check in the calling function.

I think I'm not totally understanding (or understanding at all) the synchronisation between scenarios and RNG.

Reply:
Ahhh, ok that's very clear. Wow! That sounds incredibly brittle and inefficient if we need to run part of the same stuff in both scenarios even if things are not needed.

Reply:
No kidding. And the worst thing is that I'm not convinced it actually guarantees consistency. I can imagine a scenario where an intervention reduces mortality rate, resulting in different age distributions, so different state advancement no matter how synchronised you keep the RNGs. This needs some thought, but for now we're stuck with it!